File: staticTaskAllocation/src/RAI.java
import java.util.ArrayList;
import java.util.List;
import java.util.PriorityQueue;

public class RAI {

	private List<Robot> robots;
	private List<TransportTask> transportTasks;
	private int[][] distanceMatrix;
	public COST_PERFORMANCE costPerformance;
	boolean showLogs = false;
	private int roundsOptimization = 100;

	public RAI() {
		robots = new ArrayList<Robot>();
		transportTasks = new ArrayList<TransportTask>();
		costPerformance = COST_PERFORMANCE.DISTANCE;
	}

	public void setRoundsOptimization(int rounds) {
		roundsOptimization = rounds;
	}

	public void addRobot(Robot r) {
		robots.add(r);
	}

	public void addTransportTask(TransportTask task) {
		transportTasks.add(task);
	}

	public static enum COST_PERFORMANCE {
		DISTANCE, DRIVING_TIME
	}

	public void setCostPerformance(COST_PERFORMANCE cp) {
		costPerformance = cp;
	}

	/**
	 * Creates the distance matrix
	 */
	public void createDistancMatrix() {
		int length = robots.size() + transportTasks.size();
		distanceMatrix = new int[length][length];

		// calculate distance
		for (int i = 0; i < length; i++) {
			for (int j = 0; j < length; j++) {
				// System.out.println(i + " " + j);
				distanceMatrix[i][j] = calculateDistance(i, j);
			}
		}

		System.out.println("Matrix:");
		printDistanceMatrix();
		int lowerBoundDistance = calculateLowerBoundDistance();
		System.out.println("Min. Distance: " + lowerBoundDistance);
		int lowerBoundExectionTime = calculateLowerBoundExecutionTime();
		System.out.println("Min execution Time: " + lowerBoundExectionTime);
	}

	/**
	 * Calculates the min. distanz as lower limit
	 * 
	 * @return
	 */
	private int calculateLowerBoundDistance() {
		int minValue = 0;
		int value;
		int lowerBound = 0;
		for (int i = 0; i < transportTasks.size(); i++) {
			minValue = -1;
			// search min distance to drive to each task
			for (int j = 0; j < (robots.size() + transportTasks.size() - 1); j++) {
				// search lowest value in column
				value = distanceMatrix[j][i + robots.size()];
				if (value != -1) { // Task to itself
					if (minValue == -1) {
						minValue = value;
					} else {
						if (value < minValue) {
							minValue = value;
						}
					}
				}
			}
			lowerBound += minValue;
		}
		return lowerBound;
	}

	private int calculateLowerBoundExecutionTime() {

		// ArrayList<Integer> minValues = new ArrayList<>();
		PriorityQueue<Integer> minValues = new PriorityQueue<>();

		int minValue = 0;
		int globalMinValue = -1;
		int value;
		TransportTask task;
		for (int i = 0; i < transportTasks.size(); i++) {
			minValue = -1;
			task = transportTasks.get(i);
			for (int j = 0; j < (robots.size() + transportTasks.size() - 1); j++) {
				value = distanceMatrix[j][i + robots.size()];
				value += task.getTaskDuration();
				if (value != -1) {
					if (minValue == -1) {
						minValue = value;
					} else {
						if (distanceMatrix[j][i + robots.size()] != -1) {
							if (value < minValue) {
								minValue = value;
							}
						}
					}
				}
			}
			// minValue = minValue+task.getTaskDuration();
			minValues.add(minValue);
			if (globalMinValue == -1) {
				globalMinValue = minValue;
			} else {
				if (globalMinValue > minValue) {
					globalMinValue = minValue;
				}
			}
		}

		// Is the number of tasks that at least one robots has to do together
		int minTaskCount = (int) Math.ceil((double) transportTasks.size() / (double) robots.size());

		// Take the minTaskCount values
		int lowestSum = 0;
		for (int i = 0; i < minTaskCount; i++) {
			lowestSum += minValues.poll();
		}

		return globalMinValue > lowestSum ? globalMinValue : lowestSum;
	}

	private void printDistanceMatrix() {
		int length = distanceMatrix.length;
		String line;
		for (int i = 0; i < length; i++) {
			line = "";
			for (int j = 0; j < length; j++) {
				line += distanceMatrix[i][j] + " ";
			}
			System.out.println(line);
		}
	}

	public RAIResult startAlgo() {
		int length = robots.size() + transportTasks.size();
		ArrayList<Integer> s = new ArrayList<>();

		// Start with random Node
		int rnd = getRandomNumberInRange(0, length - 1);
		s.add(rnd);
		s.add(rnd);

		while (s.size() != (length + 1)) {
			// Get next vertex
			int nextVertex = getVertexNotOnTour(s, length);

			// Insert Vertex in most cheap way in tour
			s = insertCheap(s, nextVertex);
		}
		System.out.println("First Tour: \n");
		printTour(s);

		System.out.println("Start optimization");
		ArrayList<Integer> optimizedTour = optimizeTour(s);
		System.out.println("Optimization done");
		System.out.println("Final Tour:");
		printTour(optimizedTour);

		RAIResult result = new RAIResult();
		result.deadHeadDistance = getTourDistance(optimizedTour);
		result.executionTime = getTourExecutionTime(optimizedTour);

		return result;
	}

	private ArrayList<Integer> optimizeTour(ArrayList<Integer> originalTour) {
		// 1) Choose random numbers i,j 0..(robots.size()+transportTasks.size())-1, i <=
		// j
		int max = robots.size() + transportTasks.size() - 1;
		int length = robots.size() + transportTasks.size();
		int originalTourCosts = getCostsForTour(originalTour);

		for (int counter = 0; counter < roundsOptimization; counter++) {
			print("Runde: " + counter + " Aktuelle Tour:");

			// printTour(originalTour);
			// System.out.println("");
			// System.out.println(originalTour);

			// copy tour
			ArrayList<Integer> currentTour = new ArrayList<>();
			for (int e : originalTour) {
				currentTour.add(e);
			}

			int i = getRandomNumberInRange(0, max);
			int j = getRandomNumberInRange(0, max);

			// swap if i > j
			if (i > j) {
				int tmp = i;
				i = j;
				j = tmp;
			}
			currentTour.subList(i, j + 1).clear();

			// connect vertex i-1 with vertex j + 1
			// if i = 0 und j = max, complete route deleted, delete last element
			if (i == 0 && j == max) {
				currentTour.clear();
			} else {
				// i== 0: first node was deleted, j == max: last node was deleted
				if (i == 0) {
					// delete last element
					currentTour.remove(currentTour.size() - 1);
					// Add new first node at the end of the tour to complete the tour
					currentTour.add(currentTour.get(0));
				}
			}

			// complete currentTour
			while (currentTour.size() != (length + 1)) {
				// Get next vertex
				int nextVertex = getVertexNotOnTour(currentTour, length);

				// Insert Vertex in most cheap way in tour
				currentTour = insertCheap(currentTour, nextVertex);
				checkIfTourIsValid(currentTour);
				// System.out.println(currentTour);
			}

			print("New Tour:");
			// printTour(currentTour);
			// print(currentTour);

			// compare new subPath with original Tour
			int currentTourCosts = getCostsForTour(currentTour);
			if (currentTourCosts < originalTourCosts) {
				originalTour = currentTour;
				originalTourCosts = currentTourCosts;
				System.out.println("Found Tour with lower cost (Round " + counter + "):");
				printTour(currentTour);
			} else {
				// System.out.println("old solution is better");
			}

		}
		return originalTour;
	}

	private void checkIfTourIsValid(ArrayList<Integer> tour) {
		ArrayList<Integer> tmp = new ArrayList<>();
		int counter = 0;
		for (int value : tour) {
			if (tmp.contains(value)) {
				if ((tmp.indexOf(value) == 0) && (counter == tour.size() - 1)) {

				} else {
					System.out.println("Not a valid Tour!");
				}

			} else {
				tmp.add(value);
			}
			counter++;
		}
	}

	private void printTour(ArrayList<Integer> s) {
		String ergebnis = "";
		for (int entry : s) {
			if (entry < robots.size()) {
				ergebnis += robots.get(entry).getName() + " ";
			} else {
				ergebnis += transportTasks.get(entry - robots.size()).getName() + " ";
			}
		}
		System.out.println(ergebnis + " (" + getCostsForTour(s) + ")");
	}

	/*
	 * insert the new node in the tour on position with lowest cost
	 */
	private ArrayList<Integer> insertCheap(ArrayList<Integer> tour, int element) {
		// generate tour.size-1 possible subtours

		// ArrayList<ArrayList<Integer>> list = new ArrayList<>();
		ArrayList<Integer> bestSolution = new ArrayList<>();
		int minCosts = -1;

		// case 1: tour is empty, make tour valid
		if (tour.size() == 0) {
			// System.out.println("Tour was empty!");
			bestSolution.add(element);
			bestSolution.add(element);
			return bestSolution;
		}

		// case 2: tour has 1 element, make tour valid
		if (tour.size() == 1) {
			System.out.println("Tour was not closed!");
			tour.add(element);
		}
		int rounds = tour.size() - 1;

		for (int i = 0; i < (rounds); i++) {
			if (i > 0) {
				tour.remove(i);
			}
			tour.add(i + 1, element);
			if (minCosts == -1) {
				bestSolution = tour;
				minCosts = getCostsForTour(tour);
			} else {
				int tmpDistance = getCostsForTour(tour);
				if (tmpDistance < minCosts) {
					bestSolution = tour;
					minCosts = tmpDistance;
				}
			}
		}
		return bestSolution;
	}

	/**
	 * returns the total deadhead distance to drive the tour
	 * 
	 * @param tour
	 * @return
	 */
	private int getTourDistance(ArrayList<Integer> tour) {
		int totalDistance = 0;
		for (int i = 0; i < (tour.size() - 1); i++) {
			totalDistance += distanceMatrix[tour.get(i)][tour.get(i + 1)];
		}
		return totalDistance;
	}

	private int getCostsForTour(ArrayList<Integer> tour) {
		switch (costPerformance) {
		case DISTANCE:
			return getTourDistance(tour);
		case DRIVING_TIME:
			return getTourExecutionTime(tour);
		default:
			return 0;
		}
	}

	/**
	 * Returns the total execution time for this tour
	 * 
	 * @param tour
	 * @return
	 */
	private int getTourExecutionTime(ArrayList<Integer> tour) {
		int maxExecutionTime = 0;
		int currentExecutionTime = 0;
		int executionTimeBeginning = 0;
		boolean hasSetExecutionTimeBeginning = false;
		int firstValue = tour.get(0);
		boolean firstPositionIsRobot = false;
		int lastRobot;
		// chek if first element is a robot. If not, the first elements have to be added
		// to the last robot found in the tour
		firstPositionIsRobot = checkIsRobot(tour.get(0));
		int currentValue;
		int nextValue;
		for (int i = 0; i < tour.size() - 2; i++) {
			currentValue = tour.get(i);
			nextValue = tour.get(i + 1);
			if (checkIsRobot(currentValue)) {

				// Check if tour starts with a robot. if not, if this is the first robot in the
				// tour, the execution time to this element has to be saved
				// and later added to the last robot
				if (!checkIsRobot(firstValue)) {
					if (!hasSetExecutionTimeBeginning) {
						executionTimeBeginning = currentExecutionTime;
						hasSetExecutionTimeBeginning = true;
					}
				}

				// check if current execution time > max. execution time
				if (currentExecutionTime > maxExecutionTime) {
					maxExecutionTime = currentExecutionTime;
				}

				currentExecutionTime = 0;
				lastRobot = currentValue;

				// calculate driving time from current robot to next transport task
				currentExecutionTime += distanceMatrix[currentValue][nextValue];
			} else {
				// execution time for the current transporttask
				currentExecutionTime += transportTasks.get(currentValue - robots.size()).getTaskDuration();

				// driving time from the current task to the next element (task|robot)
				currentExecutionTime += distanceMatrix[currentValue][nextValue];
			}
		}

		// check element before last. If it is not a robot, the distance from the task
		// has to be added
		// if it is a robot, the current time has to be checked
		int secondToLastValue = tour.get(tour.size() - 2);
		if (checkIsRobot(secondToLastValue)) {
			lastRobot = secondToLastValue;

			// Check if tour starts with a robot. if not, if this is the first robot in the
			// tour, the execution time to this element has to be saved
			// and later added to the last robot
			if (!checkIsRobot(firstValue)) {
				if (!hasSetExecutionTimeBeginning) {
					executionTimeBeginning = currentExecutionTime;
					hasSetExecutionTimeBeginning = true;
				}
			}

			// check if current execution time > max. execution time
			if (currentExecutionTime > maxExecutionTime) {
				maxExecutionTime = currentExecutionTime;
			}
			currentExecutionTime = 0;

		} else {
			currentExecutionTime += transportTasks.get(secondToLastValue - robots.size()).getTaskDuration();

			// check if current execution time > max. execution time
			if (currentExecutionTime > maxExecutionTime) {
				maxExecutionTime = currentExecutionTime;
			}
		}

		// check if the route starts with an robot. If not, the execution time for
		// starting tasks has to be added the the last robot
		if (!checkIsRobot(firstValue)) {
			if (executionTimeBeginning > 0) {
				currentExecutionTime += executionTimeBeginning;
				currentExecutionTime += distanceMatrix[tour.get(tour.size() - 2)][tour.get(tour.size() - 1)];

				if (currentExecutionTime > maxExecutionTime) {
					maxExecutionTime = currentExecutionTime;
				}

			}
		}
		return maxExecutionTime;
	}

	private boolean checkIsRobot(int index) {
		return index < robots.size();
	}

	/**
	 * returns a vertex that has not been added to the tour yet
	 * 
	 * @param tour
	 * @param length
	 * @return
	 */
	private int getVertexNotOnTour(ArrayList<Integer> tour, int length) {
		if (tour.size() == length + 1) {
			System.out.println("no elements left");
		}
		boolean found = false;
		int rnd;
		do {
			found = false;
			rnd = getRandomNumberInRange(0, length - 1);
			// Check if number is already used
			for (int i = 0; i < tour.size(); i++) {
				if (tour.get(i) == rnd) {
					found = true;
				}
			}
		} while (found);
		return rnd;
	}

	/**
	 * calculate distance between two elements (tasks|robots)
	 * 
	 * @param firstIndex
	 * @param secondIndex
	 * @return
	 */
	private int calculateDistance(int firstIndex, int secondIndex) {
		if (firstIndex == secondIndex) {
			return -1;
		}

		// check if elements are robots or tasks

		// Robot -> Robot
		if (firstIndex < robots.size() && secondIndex < robots.size()) {
			return 0;
		}

		// Robot -> Task
		if (firstIndex < robots.size() && secondIndex >= robots.size()) {
			Robot r1 = robots.get(firstIndex);
			TransportTask t1 = transportTasks.get(secondIndex - robots.size());
			// manhatten distance
			double erg = (Math.abs(r1.getPosition().x - t1.getPickUpPoint().x)
					+ Math.abs(r1.getPosition().y - t1.getPickUpPoint().y));
			return (int) erg;
		}

		// Task -> Task
		if (firstIndex >= robots.size() && secondIndex >= robots.size()) {
			TransportTask t1 = transportTasks.get(firstIndex - robots.size());
			TransportTask t2 = transportTasks.get(secondIndex - robots.size());
			// manhatten distance
			double erg = (Math.abs(t1.getDeliveryPoint().x - t2.getPickUpPoint().x)
					+ Math.abs(t1.getDeliveryPoint().y - t2.getPickUpPoint().y));
			return (int) erg;
		}

		// Task -> Robot
		if (firstIndex >= robots.size() && secondIndex < robots.size()) {
			return 0;
		}
		return 0;
	}

	private static int getRandomNumberInRange(int min, int max) {

		if (min >= max) {
			throw new IllegalArgumentException("max must be greater than min");
		}

		return (int) (Math.random() * ((max - min) + 1)) + min;
	}

	private void print(String s) {
		if (showLogs) {
			System.out.println(s);
		}
	}

	private void print(ArrayList<Integer> list) {
		if (showLogs) {
			System.out.println(list);
		}
	}

}

File: matlab_simulation/staticTaskAllocation/RAI.m
classdef RAI
    %RAI class calculates a costmatrix and with its help creates a tour
    %which includes all vertexes(robots and tasks) by inserting a vertex, not currently included, in the cheapest way;
    %afterwards some vertexes are removed and inserted again in the
    %cheapest way, potentially optimizing the initial tour, this is done a
    %set amount of times
    methods
        %creates matrix by calculating the cost between each vertex
        function result = createMatrix(obj, anzahlRob, anzahlAuftraege, robotList, taskList)
            for i = 1:anzahlRob + anzahlAuftraege
                for j = 1:anzahlAuftraege + anzahlRob
                    matrix(i,j) = obj.calculateMatrix(i,j, anzahlRob, robotList, taskList);    
                end
            end
            result = matrix; 
        end
        %calculates cost between two indices depending on them being robots or tasks 
        function result = calculateMatrix(~, firstIndex, secondIndex, anzahlRob, robotList, taskList)
            %if both indices are the same robot/task cost equals -1
            if eq(firstIndex, secondIndex)
                result = -1;
                return;
            end
            %if both indices are robots the costs are 0
            if (firstIndex <= anzahlRob) && (secondIndex <= anzahlRob)
                result = 0;
                return;
            end
            %the first index is a robot and the second is a task
            if (firstIndex <= anzahlRob && secondIndex > anzahlRob)
                resultx = abs(robotList(firstIndex, 2) - taskList(secondIndex - anzahlRob, 2));
                resulty = abs(robotList(firstIndex, 3) - taskList(secondIndex - anzahlRob, 3));
                result = resultx + resulty;
                return;
            end
            %both indices are tasks
            if (firstIndex > anzahlRob && secondIndex > anzahlRob)
                resultx = abs(taskList(firstIndex - anzahlRob, 2) - taskList(secondIndex - anzahlRob, 2));
                resulty = abs(taskList(firstIndex - anzahlRob, 3) - taskList(secondIndex - anzahlRob, 3));
                result = resultx + resulty; 
                return;
            end
            %if first index is a task and second is a robot the costs are 0
            if (firstIndex > anzahlRob && secondIndex <= anzahlRob)
                result = 0;
                return;
            end
        end
        
        %starting the algorithm that creates and optimizes a tour
        function result = startAlgo(obj, anzahlRob, anzahlAuftraege, matrix, costPerformance, taskList, roundsOptimization)
            %calculating lowerBound first
            lowerBoundDistance = obj.calculateLowerBoundDistance(anzahlAuftraege, anzahlRob, matrix);
            lowerBoundExecutionTime = obj.calculateLowerBoundExecutionTime(anzahlAuftraege,anzahlRob, matrix, taskList);
            
            optimizedTour = 0;
            lengthArray = anzahlRob + anzahlAuftraege;
            %choosing 2 random vertexes
            rnd(1,1)= obj.getRandomNumberInRange(1, lengthArray );
            rnd(1,2)= obj.getRandomNumberInRange(1, lengthArray );
            
            %if both are the same vertex, randomly choose a new vertex
            while rnd(1,1) == rnd(1,2)
                rnd(1,2)= obj.getRandomNumberInRange(1, lengthArray );
            end

            %gets a random vertex and includes it in the cheapest way into
            %the tour until it is complete
            while ~eq(length(rnd), lengthArray)
                nextVertex = obj.getVertexNotOnTour(rnd, lengthArray);
                currentTourInsert = obj.insertCheap(rnd, nextVertex, costPerformance,matrix, anzahlRob, taskList);
                while obj.checkIfTourIsValid(currentTourInsert) == false
                    nextVertex = obj.getVertexNotOnTour(rnd, lengthArray);      
                    currentTourInsert = obj.insertCheap(rnd, nextVertex, costPerformance,matrix, anzahlRob, taskList);
                end
                rnd = currentTourInsert;
            end
         
            %optimizes tour
            optimizedTour = obj.optimizeTour(rnd, anzahlRob, anzahlAuftraege, roundsOptimization, costPerformance, matrix, taskList);
            tour = optimizedTour;
            
            %to fully complete tour add first vertex at the end
            optimizedTour = [optimizedTour optimizedTour(1,1)];
            
            
            %calculate the distance and executionTime
            tourDistance =  obj.getTourDistance(optimizedTour ,matrix);
            tourExecutionTime =  obj.getTourExecutionTime(optimizedTour ,matrix, anzahlRob, taskList);
            
            result = [tourDistance, tourExecutionTime, tour];
        end
        
        function lowerBoundDistance = calculateLowerBoundDistance (~, anzahlAuftraege, anzahlRob, matrix)
            minValue = 0;
            value = 0;
            lowerBound = 0;

            %chooses the lowest value in matrix between 2 vertexes and adds
            %them together to create lowerBoundDistance
            for i = 1:anzahlAuftraege
                minValue = -1;
                for j = 1:anzahlRob + anzahlAuftraege
                    value = matrix(j, i+anzahlRob);
                    if value ~= -1
                        if minValue == -1
                            minValue = value;
                        elseif value < minValue
                                minValue = value;
                        end
                    end
                end
                lowerBound = lowerBound + minValue;
            end
            lowerBoundDistance = lowerBound;
        end
        function lowerBoundExecutionTime = calculateLowerBoundExecutionTime(obj, anzahlAuftraege, anzahlRob, matrix, taskList)
                minValue = 0;
                minValues = [];
                globalMinValue = -1;
                value = 0;
                
                %similar to the lowerBoundDistance the lowest value between
                %2 vertexes is chosen, the task duration is additionally
                %added
                for i=1:anzahlAuftraege
                    minValue = -1;
                    for j=1:anzahlRob
                        value = matrix(j,i);
                        value = value + obj.getTaskDuration(taskList, i);
                        if value ~= -1
                            if minValue == -1
                                minValue = value;
                            else
                                if matrix(j,i) ~= -1
                                    if value < minValue
                                        minValue = value;
                                    end
                                end
                            end
                        end
                        minValues = [minValues minValue];

                        if globalMinValue == -1
                            globalMinValue = minValue;
                        else
                            if globalMinValue > minValue
                            globalMinValue = minValue;
                            end
                        end
                    end
                end

                minTaskCount = ceil(anzahlAuftraege / anzahlRob);
                lowestSum = 0;
                for k=1:minTaskCount
                    lowestSum = lowestSum + minValues(1);
                    minValues(1) = [];
                end

                if globalMinValue > lowestSum
                    lowerBoundExecutionTime = globalMinValue;
                else
                    lowerBoundExecutionTime = lowestSum;
                end

        end 
        function result = getTaskDuration(~, taskList, i)
            result = abs(taskList(i, 2) - taskList(i,4)) + abs(taskList(i,3) - taskList(i,5));
        end

        function rnd = getRandomNumberInRange(~, min, max)
            %rnd = min + (max+min).*randi(1,1)
            rnd = randi([min max], 1);
        end   
        function result = getVertexNotOnTour(obj, rnd, lengthArray)
                found = false;
                lengthRnd = length(rnd);
                
                %chooses a random vertex that is not already found in the
                %current tour
                while true
                     found = false;
                     random = obj.getRandomNumberInRange(1, lengthArray );
                     for i=1:lengthRnd
                         if rnd(1,i) == random
                             found = true;
                         end
                     end
                    if ~found
                        break
                    end
                end
                result = random;
        end   
        function result = insertCheap (obj, rnd, nextVertex, costPerformance, matrix, anzahlRob, taskList)
            bestSolution = [];
            minCosts = -1;
            tour = rnd;
            %if the tour is empty, the current vertex is added to the
            %bestSolution
            if isempty(rnd)
                bestSolution = [bestSolution nextVertex];
                result = bestSolution;
                return;
            end
            
            %if the tour includes only one element, the current vertex is
            %added
            if length(tour) == 1
                tour = [tour nextVertex];
                result = tour;
                return;
            end

            rounds = length(tour) - 1;
            counter = 0;
            
            %for a set amount of rounds the nextVertex is included
            %inbetween all other vertexes and the solution that costs less
            %is chosen as the bestSolution
            for i=1:rounds
                if counter > 0
                    tour(i) = [];
                end
                tour = [tour(1,1:i) nextVertex tour(1,i+1:end)];
                if minCosts == -1
                    bestSolution = tour;
                    minCosts = obj.getCostsForTour(tour, costPerformance,matrix, anzahlRob, taskList);
                    %bestsolutionmincostminus = bestSolution
                end
                if minCosts ~= -1
                  tmpDistance = obj.getCostsForTour(tour, costPerformance,matrix, anzahlRob, taskList);
                  if tmpDistance < minCosts
                      bestSolution = tour;
                      minCosts = tmpDistance;
                      %bestsolutionelse = bestSolution
                  end
                end
                counter = counter +1;    
            end
            result = bestSolution;
        end
        
        %depending on the costPerformance the tour will be optimized for
        %either distance or executionTime
        function result = getCostsForTour (obj, tour, costPerformance,matrix, anzahlRob, taskList)
            if costPerformance == 0 
                result = obj.getTourDistance(tour,matrix);
                return;
            elseif costPerformance == 1
                result = obj.getTourExecutionTime(tour,matrix, anzahlRob, taskList);
                return;
            else
                result = obj.getTourDistance(tour,matrix);
            end
        end 
        function result = getTourDistance(~, tour,matrix)
            totalDistance = 0;
            %values in the matrix are added together to calculate the
            %distance
            for i=1:length(tour)-1
                totalDistance = totalDistance + matrix(tour(1,i),tour(1,i+1));
            end
            result = totalDistance;
        end
        function result = getTourExecutionTime(obj, tour, matrix, anzahlRob, taskList)
            maxExecutionTime = 0;
            currentExecutionTime = 0;
            executionTimeBeginning = 0;
            hasSetExecutionTimeBeginning = false;
            firstValue = tour(1,1);
            lastRobot = 0;

            currentValue = 0;
            nextValue = 0;

            %the calculation begins when the first robot is found
            for i=1:length(tour)-1
                currentValue = tour(1,i);
                nextValue = tour(1,i+1);
                if obj.checkIsRobot(currentValue, anzahlRob)
                    if ~obj.checkIsRobot(firstValue, anzahlRob)
                        if ~hasSetExecutionTimeBeginning
                            executionTimeBeginning = currentExecutionTime;
                            hasSetExecutionTimeBeginning = true;
                        end
                    end

                    if currentExecutionTime > maxExecutionTime
                        maxExecutionTime = currentExecutionTime;
                    end
                    currentExecutionTime = 0;
                    lastRobot = currentValue;
                    currentExecutionTime = currentExecutionTime + matrix(currentValue,nextValue);

                else
                    erg = abs(taskList(currentValue - anzahlRob, 2)-taskList(currentValue - anzahlRob, 4)) + abs(taskList(currentValue - anzahlRob, 3)- taskList(currentValue - anzahlRob, 5));
                    currentExecutionTime = currentExecutionTime + erg;

                    currentExecutionTime = currentExecutionTime + matrix(currentValue,nextValue);

                end
            end
            if length(tour) - 1 > 0
                secondToLastValue = tour(1,length(tour) - 1);
            else
                if length(tour) == 2
                secondToLastValue = tour(1,1);
                end
                if length(tour) == 1
                  secondToLastValue = tour(1,1);
                end
            end

            if obj.checkIsRobot(secondToLastValue, anzahlRob)
                lastRobot = secondToLastValue;
                if ~obj.checkIsRobot(firstValue, anzahlRob)
                    if ~hasSetExecutionTimeBeginning
                        executionTimeBeginning = currentExecutionTime;
                    end
                end

                if currentExecutionTime > maxExecutionTime
                    maxExecutionTime = currentExecutionTime;
                end
                currentExecutionTime = 0;
            else 
                 erg = abs(taskList(secondToLastValue - anzahlRob, 2)-taskList(secondToLastValue - anzahlRob, 4)) + abs(taskList(secondToLastValue - anzahlRob, 3)- taskList(secondToLastValue - anzahlRob, 5));
                 currentExecutionTime = currentExecutionTime + erg;
                if currentExecutionTime > maxExecutionTime
                    maxExecutionTime = currentExecutionTime;
                end
            end

            if ~obj.checkIsRobot(firstValue, anzahlRob)
                if executionTimeBeginning > 0
                    currentExecutionTime = currentExecutionTime + executionTimeBeginning;
                    currentExecutionTime = currentExecutionTime + matrix(tour(1,length(tour)-2), tour(1, length(tour)-1));
                    if currentExecutionTime > maxExecutionTime
                        maxExecutionTime = currentExecutionTime;
                    end
                end
            end
            result = maxExecutionTime;
        end
        function result = checkIsRobot(~, index, anzahlRob)
            result = false;
            if index <= anzahlRob
                result = true;
                return;
            end  
        end
        %tour is valid when it is not empty and a vertex is found only once
        function result = checkIfTourIsValid(~, tour)
            tmp = [];
            counter = 0;

            if length(tour) == 0
                 result = false;
                 return;
            end

            for i=tour
                if ismember(i,tmp)
                    tmpIndex = find(tmp == i);
                    if tmpIndex == 1 && counter == length(tour)-1

                    else
                        result = false;
                        return;
                    end
                else 
                    tmp = [tmp i];
                end
                counter = counter + 1;
            end
            result = true;
        end
        
        function result = optimizeTour(obj, originalTour, anzahlRob, anzahlAuftraege,roundsOptimization, costPerformance,matrix, taskList)
                max = anzahlRob + anzahlAuftraege - 1;
                Length = anzahlRob + anzahlAuftraege;
                originalTourCosts = obj.getCostsForTour(originalTour,costPerformance,matrix, anzahlRob, taskList);

                %for a set amount of times the tour is optimized
                for counter=1:roundsOptimization
                    %arraylist
                    currentTour = originalTour;

                    %2 ranom vertexes are chosen and each vertex inbetween
                    %removed
                    i = obj.getRandomNumberInRange(1, max);
                    j = obj.getRandomNumberInRange(1, max);

                    if (i > j) 
                        tmp = i;
                        i = j;
                        j = tmp;
                    end
                     x = j-i+1;
                     for f=1:x
                         currentTour(i) = [];  
                     end

                     if i == 1 && j == max
                        currentTour = [];
                     else
                        if i == 1
                            currentTour(length(currentTour)) = [];
                        end
                     end

                    currentTourInsert = [];

                    %each vertex is added again in the cheapest way
                   while (length(currentTour) ~= (anzahlRob + anzahlAuftraege))

                        nextVertex = obj.getVertexNotOnTour(currentTour, anzahlRob + anzahlAuftraege);
                        currentTourInsert = obj.insertCheap(currentTour, nextVertex, costPerformance, matrix, anzahlRob, taskList);
                        
                        while obj.checkIfTourIsValid(currentTourInsert) == false
                            obj.checkIfTourIsValid(currentTourInsert)
                            nextVertex = obj.getVertexNotOnTour(currentTour, anzahlRob + anzahlAuftraege);
                            currentTourInsert = obj.insertCheap(currentTour, nextVertex, costPerformance, matrix, anzahlRob, taskList);
                        end
                        currentTour = currentTourInsert;
                   end

                   %the costs for the new tour are compared to the costs
                   %for the initial tour, the lower value is chosen as the
                   %initial tour for the next round
                    currentTourCosts = obj.getCostsForTour(currentTour,costPerformance,matrix, anzahlRob, taskList);
                    if (currentTourCosts < originalTourCosts)
                        originalTour = currentTour;
                        originalTourCosts = currentTourCosts;
                    end
                end

                result = originalTour;
        end
    end
end

